src/sys/unix.rs

```diff
@@ -1198,6 +1304,22 @@ fn dur2linger(dur: Option<Duration>) -> libc::linger {
     }
 }
 
+/// Check that the number of buffers does not exceed the maximum for vectored IO.
+///
+/// For stream sockets, the buffer count could just be capped to the maximum.
+/// However, for datagram sockets it is not possible to split reads or writes,
+/// so we must return an error.
+fn check_iov_count(count: usize) -> io::Result<()> {
```

Comment:
Let's not return an error, but instead use `msg_iovlen: min(bufs.len(), IOV_MAX) as c_int` in `msghdr`. All the functions return the amount of bytes read/written and partial reads/writes are something callers have to deal with anyway.

Reply:
The problem with that is that for datagram sockets partial reads and writes aren't possible. You can read truncated messages, but by then the rest of the datagram is gone.

If we silently discard buffers, you could get a truncated datagram without filling the all the buffers. I don't see a sane way for user code to deal with that. As long as the whole buffer is used you can atleast check if the full buffer capacity was used.

Reply:
Unfortunately that is problem in the `recv` system call, one I don't think we should be fixing in socket2. Users using datagram socket should know this and use an appropriate sized buffer.

Reply:
I agree we don't have to fix limitations of the underlying syscalls, but I also think we should not make it worse. And giving truncated messages without using the full buffers does seem worse to me.

It would also deviating from the the syscall itself does. If the buffers exceed `IOV_MAX`, the syscall fails. Ideally, I would like to pass the `bufs.len()` unchanged to the syscall and let it report the error. But this is somewhat complicated by the fact that the `msg_iovlen` field is not a `usize` but a `socklen_t` on most platforms (except GNU/Linux and Android). So we can't pass the `usize` as is.

We could compromise and only manually return an error if `bufs.len()` exceeds the capacity of the data type of `socklen_t`, and let the OS decide how to handle it in other cases.

Reply:
Maybe we should remove the check entirely leaving it up to the user to check it. This is a low-level crate after all.

As for the truncation: it's only a problem with receiving from an UDP socket. But users of UDP should know to use buffer of ~65k bytes to ensure the entire packet can be read.

Reply:
Yeah, that is probably the most pragmatic approach.

It does leave the problem that a `usize` doesn't always fit in a `socklen_t`. I'd prefer to still return an error in that case. On GNU/Linux and Android that check can be removed, since `msg_iovlen` is a `usize`.

I admit, truncating to `socklen_t::MAX` also wouldn't be the end of the world, because it seems very unlikely that anyone would ever have more than 2^32-1 buffers. But that argument kinda goes both ways :p

Reply:
Let's just truncate to `socklen_t`, it's the simplest solution. And maybe add a note in function docs.

Reply:
Changed to code now to cap to the max for the data type of `msg_iovlen`, and let the OS handle the rest.

Also, I double-checked with `libc`, it's actually `c_int`, not `socklen_t` for the BSDs and musl. (My bad.)